src/features/annotation.ts

```diff
@@ -0,0 +1,137 @@
+
+const $annotationRenderer = Symbol('annotationRenderer');
+const $updateHotspots = Symbol('updateHotspots');
+const $hiddenAngle = Symbol('hiddenAngle');
+const $syncHiddenAngle = Symbol('syncHiddenAngle');
+
+class Hotspot extends CSS2DObject {
+  public normal: Vector3;
+
+  constructor(element: HTMLElement, position: Vector3, normal: Vector3) {
+    super(element);
+    this.position.copy(position);
+    this.normal = normal;
+  }
+}
+
+export declare interface AnnotationInterface {
+  hiddenOpacity: number;
+  hiddenAngle: string;
+  addHotspot(position: Vector3, normal: Vector3): void;
+}
+
+export const AnnotationMixin = <T extends Constructor<ModelViewerElementBase>>(
+    ModelViewerElement: T): Constructor<AnnotationInterface>&T => {
+  class AnnotationModelViewerElement extends ModelViewerElement {
+    @property({type: Number, attribute: 'hidden-opacity'})
+    hiddenOpacity: number = DEFAULT_HIDDEN_OPACITY;
+
+    @style({
+      intrinsics: {
+        basis: [numberNode(DEFAULT_HIDDEN_ANGLE, 'rad')],
+        keywords: {auto: [null]}
+      },
+      updateHandler: $syncHiddenAngle
+    })
+    @property({type: String, attribute: 'hidden-angle'})
+    hiddenAngle: string = 'auto';
+
+    private[$annotationRenderer] = new CSS2DRenderer();
+    private[$hiddenAngle] = DEFAULT_HIDDEN_ANGLE;
+
+    connectedCallback() {
+      super.connectedCallback();
+      const {style} = this[$annotationRenderer].domElement;
+      style.pointerEvents = 'none';
+      this.shadowRoot!.querySelector('.slot.default')!.appendChild(
+          this[$annotationRenderer].domElement);
+    }
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+    }
+
+    updated(changedProperties: Map<string, any>) {
+      super.updated(changedProperties);
+    }
+
+    [$syncHiddenAngle](style: EvaluatedStyle<Intrinsics<['rad']>>) {
+      this[$hiddenAngle] = style[0];
+    }
+
+    [$tick](time: number, delta: number) {
+      super[$tick](time, delta);
+      this[$updateHotspots]();
+      this[$annotationRenderer].render(this[$scene], this[$scene].activeCamera);
+    }
+
+    [$onResize](e: {width: number, height: number}) {
+      super[$onResize](e);
+      this[$annotationRenderer].setSize(e.width, e.height);
+    }
+
+    [$updateHotspots]() {
+      const {children} = this[$scene].pivot;
+      for (let i = 0, l = children.length; i < l; i++) {
+        const object = children[i];
+        if (object instanceof Hotspot) {
+          const view = this[$scene].activeCamera.position.clone();
+          view.sub(object.position);
+          const opacity = 1 -
+              (1 - DEFAULT_HIDDEN_OPACITY) *
+                  ThreeMath.smoothstep(
+                      view.angleTo(object.normal),
+                      this[$hiddenAngle] - TRANSITION_ANGLE,
+                      this[$hiddenAngle]);
+          object.element.style.opacity = `${opacity}`;
+        }
+      }
+    }
+
+    addHotspot(position: Vector3, normal: Vector3) {
```

Comment:
Why not let the user provide the DOM node?

Reply:
Sure, figured that would be the next thing. Should we provide a nice default one, or always make them do it themselves?

Reply:
One idea would be to create a `<slot>` dynamically for each added hotspot. The `<slot>` could have an opinionated default element. If the user wanted to bring their own node for the hotspot, then they would create a child of `<model-viewer>` that targets that slot. The hierarchy would look something like this:

```html
<model-viewer>
  #shadow-root
    <slot name="hotspot-1">
      <div class="default-hotspot-style"></div>
    </slot>
  <div slot="hotspot-1"><!-- anything or nothing goes here--></div>
</model-viewer>
```

Reply:
Cool, I was envisioning something along those lines. A common use case is probably having just one custom hotspot style; Can they reuse their element? Or should they be able to change our default slot CSS?

Reply:
Wouldn't it be great to have all hotspots to be part of one slot and then we could observe mutations or additions on top it, rather than making slots for each one of them? And also have a declarative api to define the hotspots .

Reply:
@lokiiarora 

> And also have a declarative api to define the hotspots .

Definitely, we're just using the imperative API as a starting point.

> Wouldn't it be great to have all hotspots to be part of **one** slot

(emphasis mine) this is also possible, but there would be a trade-off. If we use 1 slot per distributed node, then all transforms of the nodes can be opaque to the user of `<model-viewer>`. OTOH if all nodes are distributed to a single slot, we would have to transform the nodes in the light DOM, which leaks style details and risks clobbering user styles.

Reply:
> Can they reuse their element?

This is an interesting question. I think the best we could do (without delving into hacks) is let them specify a `<template>` element that contains DOM that we could "stamp" for each hotspot. There is no easy way to "instance" a DOM node in its raw form without copying it or rendering it somehow (and rendering is a huge hack with caveats).

Reply:
> Or should they be able to change our default slot CSS

Yeah, we don't need to decide this right away, but it would be easy enough to expose the styles with CSS custom properties.

Reply:
@cdata 
oh sure..
what's the advantage of putting the hotspot as a dom node rather than a 3D object?

Reply:
> what's the advantage of putting the hotspot as a dom node rather than a 3D object?

There are advantages and disadvantages.

The advantages are that a DOM node naturally has a declarative style API via CSS, participates in DOM event dispatch (`myHotspot.addEventListener('click', () => { /* custom logic */ });`) and can be replaced with any node type including custom elements.

The disadvantages include worse performance qualities and low-fidelity composition with the 3D scene. We don't get occlusion, we have to fake it. It might be hard to make this work right for complex scenes.

We aren't married to the current approach, but it's the one we want to explore first.

Reply:
Nice thought..
would love to contribute on this feature ! @cdata :)

Reply:
@cdata I've been playing with this a bit and I think I'll need a bit more specifics. The way these hotspots work is they are not children of a slot, but are referenced from the scene graph and CSS2DRenderer takes care of moving them around as needed (both in terms of position and hierarchy) to get them rendered. So I think they are not supposed to be in a slot? Only CSS2DRenderer is in a slot as of now.

Reply:
Here is the pseudo-code I'm imagining:

```javascript
class ModelViewerWithHotspots {
  addHotspot(name: string, position: Vector3, normal: Vector3) {
    const wrapper = document.createElement('div');
    wrapper.classlist.add('hotspot-slot');

    const slot = document.createElement('slot');
    slot.setAttribute('name', `hotspot-${name}`);

    const defaultHotspot = document.createElement('button');
    defaultHotspot.classlist.add('hotspot');

    wrapper.appendChild(slot);
    slot.appendChild(defaultHotspot);

    const hotspot = new Hotspot(wrapper, position, normal);
    this[$scene].pivot.add(hotspot);
  }
}
```

The hierarchy this creates looks something like this:

```html
<div class="hotspots-container">
  <div class="hotspot-slot">
    <slot name="hotspot-foo">
      <!-- The default if no other element is assigned to this slot -->
      <button class="hotspot"></button>
    </slot>
  </div>
  <div class="hotspot-slot">
    <slot name="hotspot-bar">
      <button class="hotspot"></button>
    </slot>
  </div>
</div>
```

Note that we have to use the "wrapper" div because in browsers without native ShadowDOM the `<slot>` elements will dissolve away (this detail is required to preserve the topology of the main document in such scenarios) so the wrappers are necessary evil.

So, Three.js' CSS renderer should be moving around the wrapper nodes. Since the `<button>` is a descendant of these nodes, they should be positioned correctly, and so will any custom alternatives assigned to the slot.